tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -595,9 +595,173 @@ class SPIRVEmitter : public ASTConsumer {
+  /// If a default statements exists, it becomes the body of the "else"
+  /// statement.
   void processSwitchStmtUsingIfStmts(const SwitchStmt *switchStmt) {
-    emitError("Translating Switch statements using If statements is not "
-              "implemented yet.");
+    std::vector<const Stmt *> flatSwitch;
+    flattenSwitchStmtAST(switchStmt->getBody(), &flatSwitch);
+
+    // First handle the condition variable DeclStmt if one exists.
+    // For example: handle 'int a = b' in the following:
+    // switch (int a = b) {...}
+    const auto *condVarDeclStmt = switchStmt->getConditionVariableDeclStmt();
+    if (condVarDeclStmt)
+      doStmt(condVarDeclStmt);
+
+    /// Figure out the indexes of CaseStmts (and DefaultStmt if it exists) in
+    /// the flattened switch AST.
+    /// For instance, for the following flat vector, the indexes are:
+    /// {0, 2, 5, 6, 9}
+    /// +-------------------------------------------------------------------+
+    /// |Case1|Stmt1|Case2|Stmt2|Break|Case3|Case4|Stmt4|Break|Default|Stmt5|
+    /// +-------------------------------------------------------------------+
+    std::vector<size_t> caseStmtLocs;
+    for (int i = 0; i < flatSwitch.size(); ++i)
+      if (dyn_cast<CaseStmt>(flatSwitch[i]) ||
+          dyn_cast<DefaultStmt>(flatSwitch[i]))
+        caseStmtLocs.push_back(i);
+
+    std::vector<std::shared_ptr<BinaryOperator>> comparisons;
+    std::vector<std::shared_ptr<CompoundStmt>> bodies;
+    std::shared_ptr<CompoundStmt> defaultBody;
+    std::vector<IfStmt> ifs;
+
+    // For each case, start at its index in the vector, and go forward
+    // accumulating statements until BreakStmt or end of vector is reached.
+    for (int caseNum = 0; caseNum < caseStmtLocs.size(); ++caseNum) {
+      size_t curCaseIndex = caseStmtLocs[caseNum];
+      const Stmt *curCase = flatSwitch[curCaseIndex];
+
+      // CompoundStmt to hold all statements for this case.
+      std::shared_ptr<CompoundStmt> cs =
+          std::make_shared<CompoundStmt>(CompoundStmt(Stmt::EmptyShell()));
+
+      // Accumulate all non-case/default/break statements as the body for the
+      // current case.
+      unsigned int numStatements = 0;
+      std::vector<Stmt *> statements;
+      for (int i = curCaseIndex + 1;
+           i < flatSwitch.size() && !dyn_cast<BreakStmt>(flatSwitch[i]); ++i) {
+        if (!dyn_cast<CaseStmt>(flatSwitch[i]) &&
+            !dyn_cast<DefaultStmt>(flatSwitch[i]) &&
+            !dyn_cast<BreakStmt>(flatSwitch[i])) {
+          numStatements++;
+          statements.push_back(const_cast<Stmt *>(flatSwitch[i]));
+        }
+      }
+      if (numStatements > 0)
+        cs->setStmts(astContext, statements.data(), numStatements);
+      bodies.push_back(cs);
+
+      // For non-default cases, generate the IfStmt that compares the switch
+      // value to the case value.
+      if (auto *caseStmt = dyn_cast<CaseStmt>(curCase)) {
+        IfStmt curIf = IfStmt({});
```

Comment:
Actually, there is an `ASTMaker` in the existing codebase that we can refer to to construct AST: https://github.com/Microsoft/DirectXShaderCompiler/blob/ab6759350b1f8550fa522e25154a2aa3d40e05d4/tools/clang/lib/Analysis/BodyFarm.cpp#L46